invcache/invalidating_cache.go

```diff
@@ -0,0 +1,150 @@
+package invcache
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/goware/cachestore"
+	"github.com/goware/pubsub"
+)
+
+const (
+	ChannelID = "cache_invalidation"
+)
+
+type CacheInvalidationMessage struct {
+	Key    string `json:"key"`
+	Origin string `json:"origin"`
+}
+
+type InvalidatingCache[V any] struct {
+	store      cachestore.Store[V]
+	pubsub     pubsub.PubSub[CacheInvalidationMessage]
+	instanceID string
+}
+
+func NewInvalidatingCache[V any](store cachestore.Store[V], ps pubsub.PubSub[CacheInvalidationMessage], instanceID string) *InvalidatingCache[V] {
+	return &InvalidatingCache[V]{
+		store:      store,
+		pubsub:     ps,
+		instanceID: instanceID,
+	}
+}
+
+func (ic *InvalidatingCache[V]) Exists(ctx context.Context, key string) (bool, error) {
+	return ic.store.Exists(ctx, key)
+}
+
+func (ic *InvalidatingCache[V]) Get(ctx context.Context, key string) (V, bool, error) {
+	return ic.store.Get(ctx, key)
+}
+
+func (ic *InvalidatingCache[V]) GetEx(ctx context.Context, key string) (V, *time.Duration, bool, error) {
+	return ic.store.GetEx(ctx, key)
+}
+
+func (ic *InvalidatingCache[V]) BatchGet(ctx context.Context, keys []string) ([]V, []bool, error) {
+	return ic.store.BatchGet(ctx, keys)
+}
+
+func (ic *InvalidatingCache[V]) Set(ctx context.Context, key string, value V) error {
+	if err := ic.store.Set(ctx, key, value); err != nil {
+		return err
+	}
+	return ic.publishInvalidation(ctx, key)
+}
+
+func (ic *InvalidatingCache[V]) SetEx(ctx context.Context, key string, value V, ttl time.Duration) error {
+	if err := ic.store.SetEx(ctx, key, value, ttl); err != nil {
+		return err
+	}
+	return ic.publishInvalidation(ctx, key)
+}
+
+func (ic *InvalidatingCache[V]) BatchSet(ctx context.Context, keys []string, values []V) error {
+	if err := ic.store.BatchSet(ctx, keys, values); err != nil {
+		return err
+	}
+
+	var joinedErr error
+	for _, k := range keys {
+		if err := ic.publishInvalidation(ctx, k); err != nil {
+			joinedErr = errors.Join(joinedErr, err)
+		}
+	}
+
+	return joinedErr
+}
+
+func (ic *InvalidatingCache[V]) BatchSetEx(
+	ctx context.Context,
+	keys []string,
+	values []V,
+	ttl time.Duration,
+) error {
+	if err := ic.store.BatchSetEx(ctx, keys, values, ttl); err != nil {
+		return err
+	}
+
+	var joinedErr error
+	for _, k := range keys {
+		if err := ic.publishInvalidation(ctx, k); err != nil {
```

Comment:
This may not be very efficient. It would be better to group the keys in a single message.

Reply:
agreed, publish the `keys` slice in a single message

Reply:
thanks for suggestion - done!

Reply:
I can't see this done. Maybe you forgot to push the changes?

Reply:
Done! It was misunderstanding from my side, i thought it was only about errors. And yes, great point, we reduce number of calls significantly.